src/__tests__/x509/asn1/parse.test.ts

```diff
@@ -0,0 +1,119 @@
+    expect(parseInteger(Buffer.from([0x7f]))).toEqual(BigInt(127));
+    expect(parseInteger(Buffer.from([0x00, 0x80]))).toEqual(BigInt(128));
+    expect(parseInteger(Buffer.from([0x01, 0x00]))).toEqual(BigInt(256));
+  });
+
+  it('parses negative integers', () => {
+    expect(parseInteger(Buffer.from([0xff]))).toEqual(BigInt(-1));
+    expect(parseInteger(Buffer.from([0x80]))).toEqual(BigInt(-128));
+    expect(parseInteger(Buffer.from([0xff, 0x7f]))).toEqual(BigInt(-129));
+    expect(parseInteger(Buffer.from([0xff, 0x00]))).toEqual(BigInt(-256));
+  });
+});
+
+describe('parseStringASCII', () => {
+  it('parses ASCII strings', () => {
+    expect(parseStringASCII(Buffer.from([0x48, 0x69, 0x21]))).toEqual('Hi!');
+  });
+});
+
+describe('parseTime', () => {
+  describe('with short year', () => {
+    describe('when year is 50 or greater', () => {
+      it('parses the date', () => {
+        expect(parseTime(Buffer.from('740212121110Z'), true)).toEqual(
+          new Date('1974-02-12T12:11:10Z')
+        );
+      });
+    });
+
+    describe('when year is less than 50', () => {
+      it('parses the date', () => {
+        expect(parseTime(Buffer.from('180212121110Z'), true)).toEqual(
+          new Date('2018-02-12T12:11:10Z')
+        );
+      });
+    });
+  });
+
+  describe('with long year', () => {
+    it('parses the date', () => {
+      expect(parseTime(Buffer.from('19180212121110Z'), false)).toEqual(
+        new Date('1918-02-12T12:11:10Z')
+      );
+    });
+  });
+
+  describe('when the time is invalid', () => {
+    it('throws an error', () => {
+      expect(() => parseTime(Buffer.from('FOOBAR'), true)).toThrow(
+        'invalid time'
+      );
+    });
+  });
+});
+
+describe('parseOID', () => {
+  const tests = [
+    {
+      oid: '2A8648CE3D040303',
+      expected: '1.2.840.10045.4.3.3',
+    },
+    {
+      oid: '2b0601040182371514',
+      expected: '1.3.6.1.4.1.311.21.20',
+    },
+    {
+      oid: '2a8648ce3d0201',
+      expected: '1.2.840.10045.2.1',
+    },
+    {
+      oid: '2b0601040183bf300103',
+      expected: '1.3.6.1.4.1.57264.1.3',
+    },
+    {
+      oid: '2B06010401D679020402',
+      expected: '1.3.6.1.4.1.11129.2.4.2',
+    },
+    {
+      oid: '5381ab2501',
+      expected: '2.3.21925.1',
+    },
+  ];
+
+  it('parses OID strings', () => {
+    tests.forEach(({ oid, expected }) => {
+      expect(parseOID(Buffer.from(oid, 'hex'))).toEqual(expected);
+    });
+  });
+});
+
+describe('parseBoolean', () => {
+  it('parses boolean values', () => {
+    expect(parseBoolean(Buffer.from([0x00]))).toEqual(false);
+    expect(parseBoolean(Buffer.from([0xff]))).toEqual(true);
+  });
+});
+
+describe('parseBitString', () => {
+  it('parses bit strings', () => {
+    expect(parseBitString(Buffer.from([0x04, 0x57, 0xd0]))).toEqual([
+      0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1,
+    ]);
+    expect(parseBitString(Buffer.from([0x04, 0x57, 0xdf]))).toEqual([
```

Comment:
Maybe add a comment here that the four last bits are unused and so doesn't affect the parsed value?

Reply:
👍